### PR TITLE
Feature/advanced insync discovery

### PR DESCRIPTION
**Pull request by ajanikow**

### CODE REVIEW

File: pkg/deployment/agency/state.go [hunk 1]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+func (s State) IsDBServerReadyToRestart(serverID string) bool {
+	readyToRestart := true
+
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
+		if !readyToRestart {
+			return
+		}
+
+		wc := planCollection.GetWriteConcern(0)
+		rf := planCollection.GetReplicationFactor(0)
```
Comment (ajanikow) [inline pkg/deployment/agency/state.go:162]: Can be ignored, we do not care about rf.

File: pkg/deployment/agency/state.go [hunk 2]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+
+		if wc >= len(plan) && len(plan) != 0 {
```
Comment (ajanikow) [inline pkg/deployment/agency/state.go:164]: Add special case for plan = 1

File: pkg/deployment/agency/state.go [hunk 3]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+			wc = len(plan) - 1
+		}
+
+		// Allow one server to be restarted if replication factor == write concern
+		if rf > 0 && wc == rf {
```
Comment (ajanikow) [inline pkg/deployment/agency/state.go:169]: Use above check only, this one can be skipped

File: pkg/deployment/agency/state.go [hunk 4]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+			wc -= 1
+		}
+
+		serverIsNotInSync := plan.Contains(serverID) && !current.Contains(serverID)
```
Comment (ajanikow) [inline pkg/deployment/agency/state.go:173]: Move check !plan.contains at top - ignore shard if server is not desired for this shard

File: pkg/deployment/agency/state.go [hunk 5]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+		if len(current) >= wc && serverIsNotInSync {
+			// Current shard is not in sync, but it does not matter - we have enough replicas in sync
+			// Restart of this DBServer won't affect WC
+			return
+		}
+
+		if len(current) < wc {
```
Comment (ajanikow) [inline pkg/deployment/agency/state.go:180]: < to <= - we need to what will happen if we will restart this server

File: pkg/deployment/agency/state.go [hunk 6]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
 	return nil
 }
+
+type CollectionIterator func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers)
+
+func (s State) IterateOverCollections(i CollectionIterator) {
+	for db, collections := range s.Plan.Collections {
+		for collection, details := range collections {
+			for shard, shardDetails := range details.Shards {
+				currShard := s.Current.Collections[db][collection][shard]
```
Comment (ajanikow) [inline pkg/deployment/agency/state.go:129]: Shard can be missing on current - add corner case

File: pkg/deployment/agency/state.go [hunk 7]
```diff
@@ -115,3 +119,70 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+
+func (s State) IsDBServerReadyToRestart(serverID string) bool {
+	readyToRestart := true
+
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
+		if !readyToRestart {
+			return
+		}
+
+		wc := planCollection.GetWriteConcern(0)
```
[the remark this answers is not in the document]
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go (hunk 1 above)
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go (hunk 5 above)
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go (hunk 4 above)
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go (hunk 3 above)
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go (hunk 2 above)
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go (hunk 6 above)
Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/agency/state.go [hunk 8]
```diff
@@ -115,3 +119,72 @@ func (d *StateExists) UnmarshalJSON(bytes []byte) error {
...
+		}
+	})
+
+	return isInSync
+}
+
+func (s State) IsDBServerReadyToRestart(serverID string) bool {
+	readyToRestart := true
+
+	s.IterateOverCollections(func(db, col string, planCollection *StatePlanCollection, shard string, plan ShardServers, current ShardServers) {
```
Comment (informalict) [inline pkg/deployment/agency/state.go:156]: maybe the function which is provided as an argument should return `bool` so we can stop iterating over shards earlier?

Comment (nikita-vanyasin) [reply pkg/deployment/agency/state.go]: done

File: pkg/deployment/reconcile/action_wait_for_member_in_sync.go [hunk 9]
```diff
@@ -93,7 +93,8 @@ func (a *actionWaitForMemberInSync) check() (bool, error) {
 }
 
 func (a *actionWaitForMemberInSync) checkCluster() (bool, error) {
-	if !a.actionCtx.GetShardSyncStatus() {
+	agencyState, ok := a.actionCtx.GetAgencyCache()
+	if !ok || agencyState.IsDBServerInSync(a.MemberID()) {
```
Comment (ajanikow) [inline pkg/deployment/reconcile/action_wait_for_member_in_sync.go:97]: What if you are not DBServer?

Comment (nikita-vanyasin) [reply pkg/deployment/reconcile/action_wait_for_member_in_sync.go]: Done

**Pull request merged**